### PR TITLE
Fix Astro Preview Pathing Issues

### PR DESCRIPTION
**Pull request by jonathantneal**
## Changes

- Fixes various issues triggered by `buildOptions.pageUrlFormat` and `devOptions.trailingSlash`.

## Testing

- @matthewp and I discussed `buildOptions.site` and agreed it represents a directory.
- This means a value of `http://example.com/blog` is to be treated as tho it were `http://example.com/blog/`.
- Tests are updated to reflect this view.

## Docs

bug fix only

### CODE REVIEW

File: packages/astro/src/core/preview/index.ts [hunk 1]
```diff
@@ -23,100 +22,73 @@ export interface PreviewServer {
...
-		return '/';
-	}
-	let requrl = pathname.substr(base.length);
-	return prependForwardSlash(requrl);
-}
-
 /** The primary dev action */
 export default async function preview(config: AstroConfig, { logging }: PreviewOptions): Promise<PreviewServer> {
+	// polyfill WebAPIs to globalThis for Node v12, Node v14, and Node v16
+	polyfill(globalThis, {
```
Comment (matthewp) [inline packages/astro/src/core/preview/index.ts:28]: Why is this needed?

Comment (jonathantneal) [reply packages/astro/src/core/preview/index.ts]: It was needed when I was using `.at(-1)`. I found another way to do it and removed the code.

**Review by matthewp — APPROVED** [review]

File: packages/astro/test/preview-routing.test.js [hunk 2]
```diff
@@ -207,15 +205,15 @@ describe('Preview Routing', () => {
 				expect(response.status).to.equal(404);
 			});
 
-			it('404 when loading subpath root with trailing slash', async () => {
+			it('200 when loading subpath root with trailing slash', async () => {
```
Comment (FredKSchott) [inline packages/astro/test/preview-routing.test.js:208]: this is for `trailingSlash: never` right? I imagine that this should be a 404 and that the test below should be a 200.


I think the 200->404 change below is also incorrect

**Review by FredKSchott — CHANGES_REQUESTED** [review]
requesting changes just to make sure I understand why the tests have changed, in case there was a misunderstanding of desired behavior here.  Happy to chat through this if that's the case.

File: packages/astro/test/preview-routing.test.js (hunk 2 above)
Comment (jonathantneal) [reply packages/astro/test/preview-routing.test.js:208]: This test has been updated.

**Comment by jonathantneal** [comment]
The `send` package cannot be configured like our site. We cannot alter or cancel request that it believes are valid.

https://github.com/pillarjs/send/blob/master/index.js#L728-L729

This prevents us from intercept invalid file requests. This meant the matrix of possibilities had to be written ourselves.

**Comment by matthewp** [comment]
@FredKSchott 

> requesting changes just to make sure I understand why the tests have changed, in case there was a misunderstanding of desired behavior here. Happy to chat through this if that's the case.

These tests were just added the other day by me, some were wrong.

File: packages/astro/test/preview-routing.test.js (hunk 2 above)
Comment (FredKSchott) [reply packages/astro/test/preview-routing.test.js:208]: It still seems to suffer from the problem I pointed out above: If I set `trailingSlash: 'never',` I would expect `/blog/` to return a 404. Test seemed correct as written and incorrect with this change.

Comment (FredKSchott) [reply packages/astro/test/preview-routing.test.js:208]: ah got it, `/blog/` is the subpath root and not some page within the site. Missed that in the test description, LGTM!

**Review by FredKSchott — APPROVED** [review]
